### PR TITLE
Translation of error pages

### LINKED ISSUE #1019: UI language preference ignored for kiwix-serve error pages
After #997 the text on error pages is in the language detected from the value of the "Accept-Language" header (i.e. the browser language preference), ignoring the kiwix-serve front-end UI language preference (since the information about the user-selected language is no longer sent to the server).

### PR DESCRIPTION
**Pull request by veloman-yunkan**
Will fix #1019

### CODE REVIEW

**Comment by veloman-yunkan** [comment]
@mgautierfr This PR is still work in progress, however I would like to learn your feedback regarding the small enhancement that made in mustache.hpp in order to reduce the amount of extra coding that I would need to do without it.

Conceptually, the backend part is complete (I only have to mark those error pages that should be dynamically-translatable and have KIWIX_RESPONSE_DATA in them). The frontend will only have to take advantage of the response parameter data included in the error pages and update the error page text accordingly.

**Comment by kelson42** [comment]
@mgautierfr We need a feedback here

**Comment by mgautierfr** [comment]
I'm a bit puzzled with the idea to add new specific api feature to mustache.
It would need that we ask packagers to use a patched mustache library and it would be probably (and legitmetly) refused.
We should at least propose a patch upstream (and make it accepted) before we go with this.

Why not using lambda (https://github.com/kiwix/libkiwix/issues/751#event-6459679079) ?
It would make us:
- adapt a bit the templates (to add the "call" to the translation lambda)
- Create the lambda code but it would be pretty close from what you have done in `evaluate` function (maybe even simply has you would not have to traverse data objects)

Is there any traps I haven't seen ?

---

But I really like the separation of the response from the connection. It is a step in the direction of #740 (even if this may not be the purpose)

**Comment by veloman-yunkan** [comment]
> Why not using lambda ([#751 (comment)](https://github.com/kiwix/libkiwix/issues/751#event-6459679079)) ?

@mgautierfr

Generation of the error page content is a two step process.

There is a shared template for error pages with parameters such as the page title & heading. This is the mandatory minimum needed to build an error page. However, optionally, a list of additional detail lines can be provided.

1. The error details are generated in C++ code by using parameterized message facilities. Currently (outside this PR) the translation of this additional info happens eagerly in the backend and the translated/instantiated string goes into step 2.
2. The HTML of the error page is produced by instantiating a mustache template with values for the page title, heading and details text. 

Consider it on an example:

```c++
    return HTTP404Response(request)
           + noSuchBookErrorMsg(bookName);
```

First, `noSuchBookErrorMsg(bookName)` is executed and instantiates a i18n message template for the given book name. Then the resulting `std::string` (with any explicit information about the book name erased from it) participates in another level of mustache template instantiation. The idea of this PR was to preserve all the data used to generate the response HTML and embed that data in the response so that it can be rerendered in the front-end.

This two-step nature of error response generation was missing from the `HTTPErrorResponse` unit tests but now I added a commit illustrating it.

When I started writing this reply I didn't at all see how lambdas could be used to address that problem. Now I can sense some solution relying on lambdas, however it seems artificial and limited (it won't handle additional levels involved in making up the final text; suppose that the `bookName` parameter in the example above is not a terminal string, but some text obtained as a result of translating a parameterized message: `bookName = getBookNameMsg(book, name)` ).

I will think further, but will appreciate if you hint at the solution that you had in mind if it really is capable of addressing the described challenge.

**Comment by veloman-yunkan** [comment]
I also implemented the translation of the error page in the front-end. In the worst case (if the modification to mustache.hpp is not accepted) I can switch our `ParameterizedMessage` from `kainjow::mustache::data` to a relatively small class implemented in kiwix code.

**Comment by veloman-yunkan** [comment]
I completed the draft version of the solution that doesn't need `mustache.hpp` to be patched. I will clean up the PR tomorrow.

**Comment by veloman-yunkan** [comment]
Clean-up is complete, yet there is still some work to be done on this PR. I will move forward once @mgautierfr approves the current solution. Note that to test it, you must use the last-but-one commit ("Demo of error page translation") - the extra commit was added to prove that this PR doesn't utterly break the CI.

**Review by mgautierfr — COMMENTED** [review]
Seems good to me, but I'm not sure that `Variant` is fully supported on our compilers.

We will check that and if we are good... we are good.
Else we will have to find another solution (or discuss about our compilers toolchain)

File: src/server/response.cpp [hunk 1]
```diff
@@ -151,13 +156,66 @@ std::unique_ptr<Response> Response::build_304(const ETag& etag)
...
 
-class ContentResponseBlueprint::Data : public kainjow::mustache::data
+class ContentResponseBlueprint::Data
 {
+public:
+  typedef std::list<Data>             List;
+  typedef std::map<std::string, Data> Object;
+
+private:
+  std::variant<std::string, bool, List, Object> data;
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/server/response.cpp:166]: And, sadly, this doesn't compile on aarch64 focal (https://github.com/kiwix/kiwix-build/actions/runs/7492497587/job/20396122209)

[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/response.cpp]: I provided a work-around.

[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/response.cpp]: aarch64 focal build passed https://github.com/kiwix/kiwix-build/actions/runs/7492497587/job/20454495746

[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/server/response.cpp]: Good. So now I'm happy with your proposal. Have you finish it or you still have things to do ? (In https://github.com/kiwix/libkiwix/pull/1032#issuecomment-1843007668 you said that it is no)

[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/response.cpp]: I still have to enable front-end side translation of error pages and fix the unit tests that will become broken by that change.

**Comment by veloman-yunkan** [comment]
The PR is now functionally ready, however some changes may need to be made depending on the answers to the following questions:

1. Is sending the template data with every response that may need to be translated in the frontend acceptable? That was a shortcut intended to achieve the desired functionality with the least effort and will also simplify fixing #1028.
2. Is it OK that translation data is also embedded into error responses that are not currently expected to be displayed by the viewer (e.g. responses to invalid requests to /catalog, /raw, /suggest, etc, endpoints)?

File: test/server.cpp [hunk 2]
```diff
@@ -556,11 +559,13 @@ ExpectedResponseData operator==(ExpectedResponseDataType t, std::string s)
 {
   switch (t)
   {
-    case expected_page_title: return ExpectedResponseData{s, "", "", "", ""};
-    case expected_css_url:    return ExpectedResponseData{"", s, "", "", ""};
-    case book_name:           return ExpectedResponseData{"", "", s, "", ""};
-    case book_title:          return ExpectedResponseData{"", "", "", s, ""};
-    case expected_body:       return ExpectedResponseData{"", "", "", "", s};
+    case expected_page_title: return ExpectedResponseData{s, "", "", "", "", ""};
+     case expected_css_url:   return ExpectedResponseData{"", s, "", "", "", ""};
```
Comment (mgautierfr) [inline test/server.cpp:563]: Indentation.

**Comment by mgautierfr** [comment]
> Is sending the template data with every response that may need to be translated in the frontend acceptable? That was a shortcut intended to achieve the desired functionality with the least effort and will also simplify fixing
https://github.com/kiwix/libkiwix/issues/1028.

I think we are good on this. We may in the future introduce a endpoint `error_template` to get the template from an error code (and giving this error code in the response) but for now, we can go without it. Template is a not of data, and it send only for error.

> Is it OK that translation data is also embedded into error responses that are not currently expected to be displayed by the viewer (e.g. responses to invalid requests to /catalog, /raw, /suggest, etc, endpoints)?

Yes also here. This double the error response size, but error response are pretty small and it is only for error, so pretty rare.

> Also, the test cases against vulnerabilities in kiwix-serve seem to suggest
that KIWIX_RESPONSE_DATA should be HTML-encoded too.

I totally agree.

**Comment by veloman-yunkan** [comment]
> > Also, the test cases against vulnerabilities in kiwix-serve seem to suggest
> > that KIWIX_RESPONSE_DATA should be HTML-encoded too.
> 
> I totally agree.

@mgautierfr  You better had objected in this case :smiley: . I figured out that the issue is rather an edge case and provided a simpler fix.

**Comment by mgautierfr** [comment]
> @mgautierfr You better had objected in this case 😃 . I figured out that the issue is rather an edge case and provided a simpler fix.

I had missed that, your comment opens my eyes.


---

We are good
(I've just rebase-fixup)

**Comment by veloman-yunkan** [comment]
> (I've just rebase-fixup)

Thank you. Note, however, that one of the fixup commits implied (and that was stated in its commit message) that the description of the commit-to-be-amended had to be corrected too. I have now added that information as a [commit comment](https://github.com/kiwix/libkiwix/commit/103a4516db93187b2c569b5b2a94d74f87beec96#commitcomment-137988381).

**Pull request merged**